File: 1 - Foundations/1.3 - Designing Algorithms/Code/MergeSort.js
/*
 * MergeSort.js
 *
 */

function merge(array, start, midpoint, end) {
    let lenLeft = midpoint - start + 1;
    let lenRight = end - midpoint;

    let left = new Array(lenLeft).fill(0);
    let right = new Array(lenRight).fill(0);

    for (let i = 0; i < lenLeft; i++) {
        left[i] = array[start + i];
    }

    for (let j = 0; j < lenRight; j++) {
        right[j] = array[midpoint + j + 1];
    }

    let i = 0,
        j = 0,
        k = start;

    while (i < lenLeft && j < lenRight) {
        if (left[i] <= right[j]) {
            array[k] = left[i];
            i += 1;
        } else {
            array[k] = right[j];
            j += 1;
        }
        k += 1;
    }

    while (i < lenLeft) {
        array[k] = left[i];
        i += 1;
        k += 1;
    }

    while (j < lenRight) {
        array[k] = right[j];
        j += 1;
        k += 1;
    }
}

function mergeSort(array, start, end) {
    if (start >= end) {
        return;
    }

    let midpoint = Math.floor((start + end) / 2);

    mergeSort(array, start, midpoint);
    mergeSort(array, midpoint + 1, end);

    merge(array, start, midpoint, end);
}

module.exports = mergeSort;

File: 1 - Foundations/1.3 - Designing Algorithms/Code/merge_sort.py
"""
merge_sort.py

"""

from math import floor


def merge(array, start, midpoint, end):
    """Merging procedure

    Args:
        array (list[int]): Integer array
        start (int): Start index
        midpoint (int): Midpoint index
        end (int): End index
    """
    len_left = midpoint - start + 1
    len_right = end - midpoint

    left = [0] * len_left
    right = [0] * len_right

    for i in range(0, len_left):
        left[i] = array[start + i]

    for j in range(0, len_right):
        right[j] = array[midpoint + j + 1]

    i, j, k = 0, 0, start

    while i < len_left and j < len_right:
        if left[i] <= right[j]:
            array[k] = left[i]
            i += 1
        else:
            array[k] = right[j]
            j += 1
        k += 1

    while i < len_left:
        array[k] = left[i]
        i += 1
        k += 1

    while j < len_right:
        array[k] = right[j]
        j += 1
        k += 1


def merge_sort(array, start, end):
    """Merge Sort Algorithm

    Time complexity: O(n log n)
    Space complexity: O(n)

    Args:
        array (list[int]): Integer array
        start (int): Start index
        end (int): End index
    """
    if start >= end:
        return

    midpoint = floor((start + end) / 2)

    merge_sort(array, start, midpoint)
    merge_sort(array, midpoint + 1, end)

    merge(array, start, midpoint, end)


def main():
    """
    Main method

    """
    array = [2, 4, 6, 7, 1, 2, 3, 5]
    merge_sort(array, 0, 7)
    print(array)


if __name__ == "__main__":
    main()
